### DISCUSSION TITLE
Filter roads

### DISCUSSION THREAD

**Discussion by Etheonor** [maplibre/maplibre-gl-directions#198] [Q&A]
Hello!

First of all, sorry for this (maybe) dumb question, but I'm kind of new in the OSM world and there is a lot to handle ^^ 

I created a substract of a map with Osm2pgsql and serve it with pg_tileserv. Everything's great and now I want to use maplibre-gl-directions to simulate routes. But something strange is happening, if I try to simulate a route between two point and even if the road doesn't exist in my DB, it will still go on an invisible roads.

How is that possible ? It seems to follow a correct path but I didn't generate those roads. 

![image](https://user-images.githubusercontent.com/35852770/225078724-6c5237de-0c9f-49ee-8bed-d1ab83e33819.png)

**Comment by smellyshovel** [comment]
What routing API do you use?
If it's something like OSRM, then most likely the rodes do exist in OSRM and do not in the tiles generated by you.

**Reply by Etheonor** [reply]
That's where it's funny, I don't know. 

For now, here are my steps to integrate a map to my (empty) application:

- Download an osm.pbf file
- Use osm2pgsql with "./osm2pgsql -c XXX.osm.pbf -d MYBASE -U MYUSER -W -H localhost -P 5433 -O flex -S flex-config/MYCONFIG.lua"
- Use maplibre-gl to display the map
- Use maplibre-gl directions out of the box. 

And it works...

1) Is there a default API installed?
2) Does osm2pgsql also create the routes? I don't see anything specific in my DB...

Thank you for your help

**Reply by smellyshovel** [reply]
1. Is there a default API installed?

As far as I remember, yes. OSRM is used by default.

2. Does osm2pgsql also create the routes? I don't see anything specific in my DB...

Your "local" tiles don't have *anything* to do with routing algorythms. Imagine, that if instead of displaying some real cartographic information, you displayed images of cats as tiles. Routing would still work because OSRM (or whatever other provider you decide to use) use their own info about road networks.

On the other hand, I'm very glad to hear that the plugin worked well out of the box :smile: 

> Thank you for your help

Not at all, feel free to ask if you have any other questions. Would be glad to help where I can :+1:

**Comment by Etheonor** [comment]
Ok so I can confirm OSRM is used by default, I just check with the browser debug tool and there are calls to this URL : https://router.project-osrm.org/route/v1/driving/

And so yes I agree, it works well out of the box ^^ 

Next step for me, find a way to create my own routing system (maybe use something like https://github.com/pgRouting/osm2pgrouting).

Is there anything specific you can do to tell the plugin which API to use?

**Reply by smellyshovel** [reply]
> Is there anything specific you can do to tell the plugin which API to use?

The plugin supports lots of different options. `api` is one of them: https://maplibre.org/maplibre-gl-directions/api/interfaces/MapLibreGlDirectionsConfiguration.html#api

Generally speaking, you can use any OSRM-compatible one, even write a custom one.

I worked with an API that translated graphhopper output into an OSRM-compatible format and it worked just fine. But, sadly enough, can't share any code because it's propriatary.

**Reply by Etheonor** [reply]
Almost there! Thanks again for your help :) I also asked stuff to ChatGPT, it was surprisingly useful!

I have a OSRM custom server with the routes synced on my custom osm.pbf. I configured the directions profile like that: 
`const directions = new MapLibreGlDirections(map, {
				api: 'http://localhost:5000/route/v1',
				profile: 'car'
			});`

The issue is that the url it called is not working (http://localhost:5000/route/v1/car/-4.381848808121077,48.05627689364226;-4.355022717979125,47.99962290948753? lead to a 400).

It should be the same url but without the question mark at the end

Is there something I can do on the maplibre-gl-directions configuration, or maybe it's something on the OSRM backend server to configure?

EDIT: Nevermind, I used the requestOptions property and it works! Thank again I now have a fully self hosted map with routing system :) 

![image](https://user-images.githubusercontent.com/35852770/225372319-c5682924-cf1e-4b8b-924f-5e8455562a5f.png)
